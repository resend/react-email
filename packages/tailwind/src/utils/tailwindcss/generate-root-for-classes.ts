import { parse } from 'css-tree';
import { type Config, compile } from 'tailwindcss';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import indexCss from './tailwind-stylesheets/index';
import preflightCss from './tailwind-stylesheets/preflight';
import themeCss from './tailwind-stylesheets/theme';
import utilitiesCss from './tailwind-stylesheets/utilities';

const baseCss = `
@layer theme, base, components, utilities;
@import "tailwindcss/theme.css" layer(theme);
@import "tailwindcss/utilities.css" layer(utilities);
@config;
`;

export async function generateRootForClasses(
  classes: string[],
  config: Config,
) {
  const compiler = await compile(baseCss, {
    async loadModule(id, base, resourceHint) {
      if (resourceHint === 'config') {
        return {
          path: id,
          base: base,
          module: config,
        };
      }

      throw new Error(
        `NO-OP: should we implement support for ${resourceHint}?`,
      );
    },
    async loadStylesheet(id, base) {
      if (id === 'tailwindcss') {
        return {
          base,
          path: 'tailwindcss/index.css',
          content: indexCss,
        };
      }

      if (id === 'tailwindcss/preflight.css') {
        return {
          base,
          path: id,
          content: preflightCss,
        };
      }

      if (id === 'tailwindcss/theme.css') {
        return {
          base,
          path: id,
          content: themeCss,
        };
      }

      if (id === 'tailwindcss/utilities.css') {
        return {
          base,
          path: id,
          content: utilitiesCss,
        };
      }

      throw new Error(
        'stylesheet not supported, you can only import tailwindcss',
      );
    },
  });
  const css = compiler.build(classes);
  const root = parse(css);
  resolveAllCSSVariables(root);

  return root;
}
