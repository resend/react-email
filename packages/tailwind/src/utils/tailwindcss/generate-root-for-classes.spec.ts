import { generate } from 'css-tree';
import { generateRootForClasses } from './generate-root-for-classes';

test("tailwind's generateRootForClasses()", async () => {
  expect(
    generate(
      await generateRootForClasses(
        ['text-red-500', 'sm:bg-blue-300', 'bg-slate-900'],
        {},
      ),
    ),
  ).toMatchSnapshot();
});
