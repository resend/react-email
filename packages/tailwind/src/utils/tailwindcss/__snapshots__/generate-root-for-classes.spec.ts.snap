// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tailwind's generateRootForClasses() 1`] = `"/*! tailwindcss v4.1.12 | MIT License | https://tailwindcss.com */@layer theme,base,components,utilities;@layer theme{:root,:host{}}@layer utilities{.bg-slate-900{background-color:oklch(20.8%0.042 265.755)}.text-red-500{color:oklch(63.7%0.237 25.331)}.sm\\:bg-blue-300{@media (width>=40rem){background-color:oklch(80.9%0.105 251.813)}}}"`;
